nicsdru_nidirect_theme.theme

```diff
@@ -0,0 +1,79 @@
+  ) {
+    // Only start this processing if the 'nidirect_workflow'
+    // module is installed.
+    if (\Drupal::service('module_handler')->moduleExists('nidirect_workflow')) {
+      $content_type = $variables['element']['#bundle'];
+      switch ($content_type) {
+        case 'health_condition':
+          // For health conditions, just send the user to the node edit
+          // form, where they can set the 'next review date' to a date
+          // in the future to remove this node from the 'needs audit' view.
+          $node = \Drupal::routeMatch()->getParameter('node');
+          if (!empty($node) && is_object($node)) {
+            $nid = $node->id();
+            if (!empty($nid)) {
+              // See if audit is due.
+              $next_review_date = $node->get('field_next_review_date')->value;
+              if (strtotime($next_review_date) < strtotime("now")) {
+                // Next review date is in the past,
+                // so this node is due for audit - display the audit link
+                // if the user is allowed to see it.
+                $account = User::load(\Drupal::currentUser()->id());
+                if ($account->hasPermission('audit content')) {
+                  $audit_button_text = \Drupal::config('nidirect_workflow.auditsettings')->get('audit_button_text');
+                  $audit_button_hover_text = \Drupal::config('nidirect_workflow.auditsettings')->get('audit_button_hover_text');
+                  $variables['audit_link'] = '<a href="/node/' . $nid . '/edit" title="' . $audit_button_hover_text;
+                  $variables['audit_link'] .= '" class="flag unflag-action flag-link-confirm" rel="nofollow">' . $audit_button_text . '</a>';
+                }
+              }
+            }
+          }
+          break;
+
+        case 'article':
+        case 'contact':
+        case 'page':
+          // Get the current node.
+          $node = \Drupal::routeMatch()->getParameter('node');
+          if (!empty($node) && is_object($node)) {
+            $nid = $node->id();
+            if (!empty($nid)) {
+              // See if audit is due.
+              $next_audit_date = $node->get('field_next_audit_due')->value;
+              if (strtotime($next_audit_date) < strtotime("now")) {
+                // Next audit date is in the past,
+                // so this node is due for audit - display the audit link
+                // if the user is allowed to see it.
+                $account = User::load(\Drupal::currentUser()->id());
+                if ($account->hasPermission('audit content')) {
+                  $audit_button_text = \Drupal::config('nidirect_workflow.auditsettings')->get('audit_button_text');
+                  $audit_button_hover_text = \Drupal::config('nidirect_workflow.auditsettings')->get('audit_button_hover_text');
+                  $variables['audit_link'] = '<a href="/nidirect_workflow/content_audit/' . $nid;
```

Comment:
Consider using the Link object to build these. It'll help keep things consistent with Core.

`'<a href="/nidirect_workflow/content_audit/' . $nid;` won't do any harm here as I think a) the parameter is sourced by loading the node from the routematch, and b) because I believe Twig has good XSS filtering by default... but concentating strings like this can be a red flag as it's a common XSS vector. Had this been a parameter provided from a request parameter then I think it would need a rethink on how that link is generated.

Reply:
Well it is a good thought, I would like to use the core Link class but it seems quite limited. It isn't possible to add any classes to the link or add attributes like 'rel="nofollow"'.

It is good to think about $nid and XSS, but in this case the $nid has come from $node->id() so can be fully trusted.

Reply:
Can use the Url object to see attributes:
https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Url.php/class/Url/8.2.x

Reply:
Had a look at both the Link and Url objects and finally settled on:

Link::createFromRoute

.. which seems nice and flexible and works well in this case.

Thanks both !